File: julia/src/Simulator.jl
# File: Simulator.jl
# Project: src
# Created Date: 10/10/2022
# Author: Shun Suzuki
# -----
# Last Modified: 28/04/2023
# Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
# -----
# Copyright (c) 2022 Shun Suzuki. All rights reserved.
# 


mutable struct Simulator
    _builder::Ptr{Cvoid}
    log_level
    log_func
    timeout
    build
    function Simulator()
        chandle = Ref(Ptr{Cvoid}(0))
        autd3capi_link_simulator.autd_link_simulator(chandle)
        simulator = new(chandle[])
        simulator.log_level = function (level::UInt8)
            autd3capi_link_simulator.autd_link_simulator_log_level(simulator._builder, level)
            simulator
        end
        simulator.log_func = function (out::Function, flush::Function)
            ffout = (x::Cstring) -> out(x)
            ffflush = () -> flush()
            pout = @cfunction($ffout, Cvoid, (Cstring,))
            pflush = @cfunction($ffflush, Cvoid, ())
            autd3capi_link_simulator.autd_link_simulator_log_func(simulator._builder, pout, pflush)
            simulator
        end
        simulator.timeout = function (timeout::UInt64)
            autd3capi_link_simulator.autd_link_simulator_timeout(simulator._builder, timeout)
            simulator
        end
        simulator.build = function ()
            chandle = Ref(Ptr{Cvoid}(0))
            autd3capi_link_simulator.autd_link_simulator_build(chandle, simulator._builder)
            Link(chandle[])
        end
        simulator
    end
end

File: matlab/simulator.m
%{
%File: simulator.m
%Project: autd3-matlab
%Created Date: 11/06/2022
%Author: Shun Suzuki
%-----
%Last Modified: 02/02/2023
%Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
%-----
%Copyright (c) 2022 Shun Suzuki. All rights reserved.
%
%}

addpath('autd3')

Error = [];

use_link_soem = false;
use_backend_cuda = false;

try
    init_autd(use_link_soem, use_backend_cuda);

    builder = GeometryBuilder();
    builder.add_device([0 0 0], [0 0 0]);
    geometry = builder.build();

    sim = Simulator();
    link = sim.build();

    cnt = Controller(geometry, link);

    runner(cnt);
catch Error

    for e = Error
        disp(e);
    end

end

deinit_autd(use_link_soem, use_backend_cuda);
